webui/react/src/pages/ExperimentDetails/ExperimentVisualization/HpScatterPlots.tsx

```diff
@@ -0,0 +1,239 @@
+import { Alert, Select } from 'antd';
+import { SelectValue } from 'antd/es/select';
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+
+import Grid, { GridMode } from 'components/Grid';
+import Message, { MessageType } from 'components/Message';
+import MetricBadgeTag from 'components/MetricBadgeTag';
+import MetricSelectFilter from 'components/MetricSelectFilter';
+import MultiSelect from 'components/MultiSelect';
+import ResponsiveFilters from 'components/ResponsiveFilters';
+import ScatterPlot from 'components/ScatterPlot';
+import Section from 'components/Section';
+import SelectFilter from 'components/SelectFilter';
+import Spinner from 'components/Spinner';
+import useResize from 'hooks/useResize';
+import { V1TrialsSnapshotResponse } from 'services/api-ts-sdk';
+import { detApi } from 'services/apiConfig';
+import { consumeStream } from 'services/utils';
+import { ExperimentBase, ExperimentHyperParamType, MetricName, metricTypeParamMap } from 'types';
+import { isNumber } from 'utils/data';
+import { metricNameToStr } from 'utils/string';
+import { terminalRunStates } from 'utils/types';
+
+import css from './HpScatterPlots.module.scss';
+
+const { Option } = Select;
+
+interface Props {
+  batches: number[];
+  experiment: ExperimentBase;
+  hParams: string[];
+  isLoading?: boolean;
+  metrics: MetricName[];
+  onBatchChange?: (batch: number) => void;
+  onHParamChange?: (hParams?: string[]) => void;
+  onMetricChange?: (metric: MetricName) => void;
+  selectedBatch: number;
+  selectedHParams: string[];
+  selectedMetric: MetricName;
+}
+
+interface HpMetricData {
+  hpLogScales: Record<string, boolean>;
+  hpValues: Record<string, number[]>;
+  metricValues: Record<string, number[]>;
+  trialIds: number[];
+}
+
+const ScatterPlots: React.FC<Props> = ({
+  batches,
+  experiment,
+  hParams,
+  isLoading = false,
+  metrics,
+  onBatchChange,
+  onHParamChange,
+  onMetricChange,
+  selectedBatch,
+  selectedHParams,
+  selectedMetric,
+}: Props) => {
+  const baseRef = useRef<HTMLDivElement>(null);
+  const [ hasLoaded, setHasLoaded ] = useState(false);
+  const [ chartData, setChartData ] = useState<HpMetricData>();
+  const [ pageError, setPageError ] = useState<Error>();
+  const resize = useResize(baseRef);
```

Comment:
ultra-nit: in general I tend to put breaklines between interdependent vars, like in that case `resize` depends on `baseRef`. Are rules like this enforced in the codebase as good practices? Or there are others? Or I'm overlooking a super-minor thing?

Reply:
I think that's a good rule, it's just typically hard to enforce if it's not automatic. But I'll do that going forward. Stuff like this would be great to discuss in WebUI AG. to get everyone on the same page.